[After #401] Transpile with pass managers

beckykd opened the pull request:
Closes #211

javabster commented on docs/transpile/custom_passes.mdx:
````diff
@@ -0,0 +1,101 @@
+```
+![Description](/images/transpile/custom_passes/qc_dd_basis.png)
+
+We can see the absence of `YGate`s in our circuit, and we have explicit timing information in the form of `Delay` gates. This transpiled circuit with dynamical decoupling can now be sent to the backend.
````
Just to round off this example could you add a code snippet showing how you'd send this circuit on to a backend?

kevinsung replied:
Wouldn't that violate the separate nature of the "run" section?

kevinsung replied:
We decided to simply link to the run section.

javabster commented on docs/transpile/custom_passes.mdx:
```diff
@@ -0,0 +1,101 @@
+description: Passes not included in the transpiler default optimization levels.
+
+---
+# Custom passes 
```
Is it correct to call this section "Custom Passes"? To me "custom passes" implies you're creating the passes yourself, but really you are taking existing passes from qiskit but putting them together in a custom passmanager. So I would have thought a better title for this would be "custom passmanagers". But I admit this section is a bit beyond my expertise so if you think "custom passes" is the correct terminology we can stick with that

kevinsung replied:
I renamed it to "Customizing transpilation"

javabster commented on docs/transpile/custom_passes.mdx:
```diff
@@ -0,0 +1,101 @@
+---
+# Custom passes 
+
+Many transpiler passes within Qiskit are optional and are not executed in any of the default optimization levels. In some cases, they can take the place of default passes within those stages. Others can be written outside of Qiskit and executed as plugins. Here we introduce a pass, `PadDynamicalDecoupling`, that adds an error suppression technique called _dynamical decoupling_ to the circuit. Bec ...
```
I like the first 2 sentences of the intro but after that it wasn't clear to me when you say youre "introducing" a pass whether its one that's already available in qiskit or one you are making from scratch. I managed to guess later on by looking at the imports lol but it would be good to make that clearer in the intro paragraph. I'd recommend instead saying something like "In this page we'll be constructing a custom passmanager including built-in Qiskit passes PadDynamicalDecoupling, ALAPScheduleAanalysis and BasisTranslator. This does dynamical decoupling blah blah..."

kevinsung replied:
I rewrote this.

javabster commented on docs/transpile/custom_passes.mdx:
````diff
@@ -0,0 +1,101 @@
+
+```python
+    qc = EfficientSU2(12, entanglement='circular', reps=1)
+    qc_t = transpile(qc, target=target, seed_transpiler=12345)
````
General question - why is it necessary to run `transpile()` when you're going to be running your custom passes later on? Is there a specific reason to do 2 rounds of transpilation? I'm guessing im not the only one who would have this question so might be nice to add some explanation into the text itself :D

kevinsung replied:
I added a sentence of explanation:

> We'll first transpile the circuit to the backend because dynamical decoupling pulses need to be added after the circuit has been transpiled and scheduled.

javabster commented on docs/transpile/custom_passes.mdx:
````diff
@@ -0,0 +1,101 @@
+
+Next, we execute the custom passes. We instantiate the `PassManager` with `ALAPScheduleAnalysis` and `PadDynamicalDecoupling`.  We must run `ALAPScheduleAnalysis` first to add timing information about the quantum circuit before the regularly-spaced dynamical decoupliong sequences can be added. These passes are run on the circuit with `.run()`.
+
+```python
+    pm = PassManager([
+        ALAPScheduleAnalysis(target=target),
+        PadDynamicalDecoupling(target=target, dd_sequence=dd_sequence),
+        ])
+    qc_dd = pm.run(qc_t)
+```
````
I feel like this code block is the most important part of this page, I think it would be good to emphasise it somehow and make the point to the reader that this is the basic structure for constructing and running a custom passmanager, and where they could go to look at the other qiskit passes that are available. 

Possibly the page introduction should start with a code block like this (maybe just pseudo code) as a way to explain in general how to construct/run the PassManager object, and highlight the different passes available. Then launch into the worked example for dynamical decoupling.

kevinsung replied:
I rewrote the whole page.

javabster reviewed (changes requested):
Looking good so far!

kevinsung commented:
I just noticed that the file `defaults_and_configuration_options.mdx` is still a WIP:
>** ~~~Question about this: "approximation_degree (float) - Heuristic dial used for circuit approximation (1.0=no >approximation, 0.0=maximal approximation). The default value is 1.0."
>
>It looks like the value is a float in the range [0 - 1]. Should we clarify that in the description?  What do the "in between" >measurements mean?**

 @beckykd Is this the only remaining issue that needs to be addressed?

javabster commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,414 @@
+{
```
<blockquote>Qiskit includes functions to build preset <code><a href="https://github.com/api/qiskit/qiskit.transpiler.PassManager" target="_blank">PassManager</a></code> objects.&nbsp;These preset passmanagers are used by the <code><a href="https://github.com/api/qiskit/qiskit.utils.QuantumInstance#transpile" target="_blank">transpile</a></code> function&nbsp;for each optimization level. There are 4 optimization levels ranging from 0 to 3, where higher&nbsp;optimization levels take more time and computational effort but may yield a&nbsp;more optimal circuit.&nbsp;Optimization level 0 is intended for device characterization experiments and, as such, only&nbsp;maps the input circuit to the constraints of the target backend, without&nbsp;performing any optimizations. Optimization level 3 spends the most effort to optimize the circuit.&nbsp;However, as many of the optimization techniques in the transpiler are heuristic based, spending more&nbsp;computational effort does not always result in an improvement in the quality of the output&nbsp;circuit.</blockquote><p>I think this info is already covered in another page, so you can probably cut it down or remove it entirely. For this intro I think it would be better to indicate why the reader might want to work directly with preset passmanagers in the first place, and give an outline of what will be covered in the whole page</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Qiskit/documentation/pull/391/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

kevinsung replied:
Done.

javabster commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,414 @@
+{
```
<p>in this code example could you also include how to run the passmanger? and update the description above accordingly</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Qiskit/documentation/pull/391/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

kevinsung replied:
Done.

javabster commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,414 @@
+{
```
<p>Is there a reference we can add here for people to learn more about dynamical decoupling?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Qiskit/documentation/pull/391/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

javabster replied:
<p>oh I realise now there is a referene for dd haha. Could you move this up to the intro paragraph where you first mention it? I think it would also be good to put it in and admonition</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Qiskit/documentation/pull/391/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

javabster commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,414 @@
+{
```
<p>could you explain here what the difference is between a regular pass manager and a staged pass manager?</p><p> </p><p>Also, I feel like maybe this section should go above the "create your own passmanager section", what do you think? This part seems significantly easier than the creating your own passmanger section, and I think it's generally best to order the page with easiest stuff first</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Qiskit/documentation/pull/391/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

kevinsung replied:
Done.

javabster commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,414 @@
+{
```
<p>could you add a "Next steps" box to the end like we have for the other docs pages please?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Qiskit/documentation/pull/391/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

kevinsung replied:
Done.

kevinsung commented:
@javabster I've addressed your comments. In doing so, I added some links to the still-non-existent page on "write your own transpiler pass" which I believe will be added in #401

javabster commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,468 @@
+   "source": [
+    "## Creating your own staged pass manager\n",
+    "\n",
+    "A `StagedPassManager` is a pass manager that is composed of individual stages, where each stage is defined by a PassManager instance. You can create a `StagedPassManager` simply by specifying the desired stages. For example, the following code creates a staged pass manager with two stages, `init` and `translation`. The `translation` stage is set to the pass manager that we created previously ...
```
> The `translation` stage is set to the pass manager that we created previously.
What does this mean?

javabster commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,468 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAABI4AAANoCAYAAACm0aGIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/SrBM8AAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdd3gU5doG8Hu2phcSSgglEQhNCdJBUJAmvQlKLwKKWOB4UNGjJ6AiKBzpIPphFBCkdwSUroA0QZQuRCAk1CSkbLK7M98fSxaW3ZmQzSa7m9y/6+LSzLM7+86TyTszz77zjiBJkgQiIiIiIiIiIqKHqNzdACIiIiIiIiIi8kwsHBERERERERERkUMsHBERERERERERkUM ...
```
I can't tell if this is just an issue with the notebook review display, but the "Next steps" section should be inside an admonition like it is on the rest of the pages

javabster commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,468 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Customizing transpilation\n",
```
```suggestion
    "# Create custom pass managers\n",
```

javabster commented on docs/transpile/_toc.json:
```diff
@@ -1,18 +1,21 @@
+      "url": "/transpile/defaults_and_configuration_options"
+    },
+    {
+      "title": "Customizing transpilation",
```
```suggestion
      "title": "Create custom pass managers",
```

javabster commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,470 @@
+   "source": [
+    "## Creating your own staged pass manager\n",
+    "\n",
+    "A `StagedPassManager` is a pass manager that is composed of individual stages, where each stage is defined by a PassManager instance. You can create a `StagedPassManager` simply by specifying the desired stages. For example, the following code creates a staged pass manager with two stages, `init` and `translation`. The `translation` stage is set to the pass manager that we created previously ...
```
```suggestion
    "A `StagedPassManager` is a pass manager that is composed of individual stages, where each stage is defined by a `PassManager` instance. You can create a `StagedPassManager` simply by specifying the desired stages. For example, the following code creates a staged pass manager with two stages, `init` and `translation`. The `translation` stage is defined by the pass manager that we created previously."
```

kevinsung commented:
This is blocked by #401 because we include links to that page.

abbycross reviewed (approved):
some small suggestions, looks good otherwise!

kevinsung commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,470 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAABI4AAANoCAYAAACm0aGIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/SrBM8AAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdd3gU5doG8Hu2phcSSgglEQhNCdJBUJAmvQlKLwKKWOB4UNGjJ6AiKBzpIPphFBCkdwSUroA0QZQuRCAk1CSkbLK7M98fSxaW3ZmQzSa7m9y/6+LSzLM7+86TyTszz77zjiBJkgQiIiIiIiIiIqKHqNzdACIiIiIiIiIi8kwsHBERERERERERkUMsHBERERERERERkUM ...
```
Done.

kevinsung commented on docs/transpile/customizing-transpilation.ipynb:
```diff
@@ -0,0 +1,470 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAABI4AAANoCAYAAACm0aGIAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/SrBM8AAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdd3gU5doG8Hu2phcSSgglEQhNCdJBUJAmvQlKLwKKWOB4UNGjJ6AiKBzpIPphFBCkdwSUroA0QZQuRCAk1CSkbLK7M98fSxaW3ZmQzSa7m9y/6+LSzLM7+86TyTszz77zjiBJkgQiIiIiIiIiIqKHqNzdACIiIiIiIiIi8kwsHBERERERERERkUMsHBERERERERERkUM ...
```
It looks like you already suggested this.

abbycross replied:
Weird. Github definitely started freaking out on me so I deleted my suggested changes here. It seemed to think I was suggesting to replace an entire image even though the image wasn't in the line in question. As long as no images have been deleted, I'm fine 🤷

frankharkins commented on docs/transpile/custom-pass-managers.ipynb:
```diff
@@ -0,0 +1,470 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The [stage generator functions](/api/qiskit/transpiler_preset#stage-generator-functions) functions may be useful for the construction of custom pass managers.\n",
```
```suggestion
    "The [stage generator functions](/api/qiskit/transpiler_preset#stage-generator-functions) may be useful for the construction of custom pass managers.\n",
```

frankharkins reviewed (commented):
Looks good! I learnt a lot reading this 🙂

beckykd commented:
> I just noticed that the file `defaults_and_configuration_options.mdx` is still a WIP:
> 
> > ** ~~~Question about this: "approximation_degree (float) - Heuristic dial used for circuit approximation (1.0=no >approximation, 0.0=maximal approximation). The default value is 1.0."
> > It looks like the value is a float in the range [0 - 1]. Should we clarify that in the description?  What do the "in between" >measurements mean?**
> 
> @beckykd Is this the only remaining issue that needs to be addressed?

There's a lot more information about this setting in the file common-paramaters.mdx, but I'll add some clarification in this file as well, and a link to the common-parameters file.

kevinsung commented:
@beckykd I deleted that file from this PR because I thought it was being worked on at https://github.com/Qiskit/documentation/pull/267.

beckykd commented:
> @beckykd I deleted that file from this PR because I thought it was being worked on at #267.

267 has been merged now, so you can make changes to it via any PR and not cause merge conflicts.  I just thought it would be easier to make the change here since it's already open and being worked.

kevinsung commented:
I see. Ok, it looks good. Thanks!

beckykd commented on docs/transpile/custom-pass-managers.ipynb:
```diff
@@ -0,0 +1,470 @@
+   "metadata": {},
+   "source": [
+    "# Create custom pass managers\n",
+    "\n",
```
Can we add an intro here, explaining what a pass manager is, including why you would use one?

beckykd commented:
I'll add the link in after everything is merged.  This work is tracked here: https://github.com/Qiskit/documentation/issues/378

kevinsung commented:
@javabster @beckykd I've rewritten the introduction to
- Highlight pass managers as the recommended way to perform transpilation
- Add an admonition telling users to set `skip_transpilation=True` if they transpile locally

I've also updated the notebook
- Added a section called "What is a (staged) pass manager?" right after the intro paragraph.
- Renamed it to "Transpile with pass managers" because it's now a general overview of pass managers.

Please take another look.

abbycross commented on docs/transpile/defaults-and-configuration-options.mdx:
```diff
@@ -65,6 +65,8 @@ All parameters except `circuits` are optional.  For full details, see the [Trans
 
 These arguments describe the system that the input circuit will be run on.  For example, they may be used to ensure that the returned circuit can be run on the specified backend, where the transpiler maps the circuit's virtual qubits to the physical qubits which may have limited connectivity as specified in the `coupling_map`, `basis_gates` that specify which single- and two-qubit gates are calib ...
 
+Many of these parameters are described in detail in [Commonly used parameters](common-parameters).
```
```suggestion
Many of these parameters are described in detail in [Other commonly used parameters](common-parameters).
```

kevinsung replied:
There's a mismatch between the page title and its toc entry. What should it be?

beckykd replied:
I think they should both be "Commonly..."  @javabster ?

javabster replied:
I deliberately made them different. My logic was that when a user is looking at the sidebar it makes sense to use the word "other" in the context of the other page titles. But if a user just goes directly to the page (e.g. from google search) and isn't paying attention to the other sidebar headings then just reading the page title with the word "other" in it might be coinfusing

javabster replied:
I don't have a strong feeling though, if you think its better to keep it the same thats fine by me

kevinsung replied:
I made the names the same.

beckykd commented on docs/transpile/transpile-with-pass-managers.ipynb:
```diff
@@ -0,0 +1,483 @@
+   "source": [
+    "## What is a (staged) pass manager?\n",
+    "\n",
+    "In the context of Qiskit, transpilation refers to the process of transforming an input circuit into a form that is suitable to be executed on a quantum device. Transpilation typically occurs in a sequence of steps called transpiler passes. The circuit is processed by each transpiler pass in sequence, with the output of one pass becoming the input to the next. For example, one pass could go t ...
```
```suggestion
    "In the context of Qiskit, transpilation refers to the process of transforming an input circuit into a form that is suitable for execution on a quantum device. Transpilation typically occurs in a sequence of steps called transpiler passes. The circuit is processed by each transpiler pass in sequence, with the output of one pass becoming the input to the next. For example, one pass could go through the circuit and merge all consecutive sequences of single-qubit gates, and then the next pass could synthesize these gates into the basis set of the target device. The transpiler passes included with Qiskit are located in the [qiskit.transpiler.passes](/api/qiskit/transpiler_passes) module.\n",
```

beckykd commented on docs/transpile/transpile-with-pass-managers.ipynb:
```diff
@@ -0,0 +1,483 @@
+    "\n",
+    "In the context of Qiskit, transpilation refers to the process of transforming an input circuit into a form that is suitable to be executed on a quantum device. Transpilation typically occurs in a sequence of steps called transpiler passes. The circuit is processed by each transpiler pass in sequence, with the output of one pass becoming the input to the next. For example, one pass could go t ...
+    "\n",
+    "A pass manager is an object that stores a list of transpiler passes and can execute them on a circuit. You create a pass manager by initializing a [PassManager](/api/qiskit/qiskit.transpiler.PassManager) with your list of transpiler passes. To run the transpilation on a circuit, call the [run](/api/qiskit/qiskit.transpiler.PassManager#run) method with your circuit as input.\n",
```
```suggestion
    "A pass manager is an object that stores a list of transpiler passes and can execute them on a circuit. Create a pass manager by initializing a [PassManager](/api/qiskit/qiskit.transpiler.PassManager) with a list of transpiler passes. To run the transpilation on a circuit, call the [run](/api/qiskit/qiskit.transpiler.PassManager#run) method with a circuit as input.\n",
```

beckykd commented on docs/transpile/transpile-with-pass-managers.ipynb:
```diff
@@ -0,0 +1,483 @@
+    "\n",
+    "A pass manager is an object that stores a list of transpiler passes and can execute them on a circuit. You create a pass manager by initializing a [PassManager](/api/qiskit/qiskit.transpiler.PassManager) with your list of transpiler passes. To run the transpilation on a circuit, call the [run](/api/qiskit/qiskit.transpiler.PassManager#run) method with your circuit as input.\n",
+    "\n",
+    "A staged pass manager is a special kind of pass manager that represents a level of abstraction above that of a normal pass manager. While a normal pass manager is composed of a number of transpiler passes, a staged pass manager is composed of a number of *pass managers*. This is a useful abstraction because transpilation is typically happens in discrete stages, as described in [Transpiler st ...
```
```suggestion
    "A staged pass manager is a special kind of pass manager that represents a level of abstraction above that of a normal pass manager. While a normal pass manager is composed of several transpiler passes, a staged pass manager is composed of several *pass managers*. This is a useful abstraction because transpilation typically happens in discrete stages, as described in [Transpiler stages](transpiler-stages), with each stage being represented by a pass manager. The rest of this page describes how to create and customize (staged) pass managers."
```

beckykd commented on docs/transpile/transpile-with-pass-managers.ipynb:
```diff
@@ -0,0 +1,483 @@
+   "source": [
+    "See [Transpilation defaults and configuration options](defaults-and-configuration-options) for a description of the possible arguments to the `generate_preset_pass_manager` function.\n",
+    "\n",
+    "If the preset pass managers don't fulfill your needs, you can customize transpilation by creating your own (staged) pass managers or even your own transpilation passes. The rest of this page describes how to create pass managers. For instructions to create transpilation passes, see [Write your own transpilation pass](write-your-own-transpilation-pass)."
```
```suggestion
    "If the preset pass managers don't fulfill your needs, customize transpilation by creating (staged) pass managers or even transpilation passes. The rest of this page describes how to create pass managers. For instructions to create transpilation passes, see [Write your own transpilation pass](write-your-own-transpilation-pass)."
```

beckykd commented on docs/transpile/transpile-with-pass-managers.ipynb:
```diff
@@ -0,0 +1,483 @@
+   "source": [
+    "## Create your own pass manager\n",
+    "\n",
+    "The [qiskit.transpiler.passes](/api/qiskit/transpiler_passes) module includes many transpiler passes that you can use to create pass managers. To create a pass manager, initialize a `PassManager` with a list of passes. For example, the following code creates a transpiler pass that merges adjacent two-qubit gates and then synthesizes them into a basis of [$R_y$](/api/qiskit/qiskit.circuit.lib ...
```
```suggestion
    "The [qiskit.transpiler.passes](/api/qiskit/transpiler_passes) module includes many transpiler passes that can be used to create pass managers. To create a pass manager, initialize a `PassManager` with a list of passes. For example, the following code creates a transpiler pass that merges adjacent two-qubit gates and then synthesizes them into a basis of [$R_y$](/api/qiskit/qiskit.circuit.library.RYGate), [$R_z$](/api/qiskit/qiskit.circuit.library.RZGate), and [$R_{xx}$](/api/qiskit/qiskit.circuit.library.RXXGate),  gates."
```

beckykd commented:
> @javabster @beckykd I've rewritten the introduction to
> 
> * Highlight pass managers as the recommended way to perform transpilation
> * Add an admonition telling users to set `skip_transpilation=True` if they transpile locally
> 
> I've also updated the notebook
> 
> * Added a section called "What is a (staged) pass manager?" right after the intro paragraph.
> * Renamed it to "Transpile with pass managers" because it's now a general overview of pass managers.
> 
> Please take another look.

The introduction information is excellent!  I feel like I have a much better understanding of how they all fit together now.

javabster commented on docs/transpile/transpile-with-pass-managers.ipynb:
```diff
@@ -0,0 +1,483 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "See [Transpilation defaults and configuration options](defaults-and-configuration-options) for a description of the possible arguments to the `generate_preset_pass_manager` function.\n",
```
Does `generate_preset_pass_manager` take all the same inputs as `transpile()`? If yes we should mention that explicitly

kevinsung replied:
I'll update the "Transpilation default settings and configuration options" in this PR.

kevinsung replied:
Never mind, I realized that page doesn't need to be changed. I mentioned that the arguments match.

javabster reviewed (commented):
Phenomenal work kevin 👏

Additional question, could be addressed here or in a followup PR: do you think we should update the "Transpiler Stages" page as well? Atm it leads with a `transpile()` narrative, but is there a way to refactor it to fit in better with this passsmanager narrative, like you did for the "Introduction" page?

Additional Additional question: do you think the "Create a dd passmanager" section should be a page on it's own? It feels like a pretty high profile example, maybe it warrants more attention with its own page?

kevinsung commented:
>do you think we should update the "Transpiler Stages" page as well? Atm it leads with a transpile() narrative, but is there a way to refactor it to fit in better with this passsmanager narrative, like you did for the "Introduction" page?

Done in this PR.

javabster commented on docs/transpile/transpiler-stages.ipynb:
```diff
@@ -4,18 +4,20 @@
+    "This page describes the stages of Qiskit's prebuilt transpilation pipeline.\n",
     "\n",
-    "Qiskit preset `PassManager` objects are objects that manage a set of different transpiler `Passes`. When the `transpile()` function is called, a preset `StagedPassManager` object is created and then executed. Depending on how the `optimization_level` argument for `transpile()` is set, one of four different preset pass managers is used, ranging from very light to heavy optimization (for more  ...
+    "Create a preset staged pass manager using the [`generate_preset_pass_manager`](/api/qiskit/transpiler_preset#qiskit.transpiler.preset_passmanagers.generate_preset_pass_manager) function. This function has one positional argument that must be specified, the `optimization_level`, which is an integer that can be 0, 1, 2, or 3. Higher values indicate heavier but more costly optimization (see [Tr ...
```
Sorry I should have been clearer, I don't think we should completely remove the paragraph about the transpile function, I just meant we should _also_ talk about how these stages relate to people's workflow when they're not using `transpile()`

kevinsung replied:
I rewrote it.

javabster commented on docs/transpile/transpiler-stages.ipynb:
```diff
@@ -4,18 +4,20 @@
-    "1. `translation`\n",
-    "1. `optimization`\n",
-    "1. `scheduling`"
+    "2. `layout`\n",
```
I think the multiple `1.` was intentional, bc markdown is clever enough to figure out the numbering so you don't have to

kevinsung replied:
Done.

javabster reviewed (approved):
LGTM 🚀
